File: Python/day04/AOC.py
# Original solution invloved finding subgrids of 4x4 and checking if they contained the word "XMAS".
# That solution resulted in a lot of double counting and got really messy.
# I then realized that I could just check for the word "XMAS" in all 8 directions from each "X" and that would be enough.
def part1(input):
    count = 0
    dirs = [(-1, 0), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1), (-1, 1), (0, 1)]
    rows, cols = len(input), len(input[0])
    for row in range(rows):
        for col in range(cols):
            if input[row][col] != "X":
                continue
            for dr, dc in dirs:
                if 0 <= row + 3 * dr < rows and 0 <= col + 3 * dc < cols:
                    if "".join([input[row + i*dr][col + i*dc] for i in range(4)]) == "XMAS":
                        count += 1
    return count

def part2(input):
    count = 0
    rows, cols = len(input), len(input[0])
    valid_corners = {"MMSS", "MSSM", "SSMM", "SMMS"} # all possible corner combinations if read clockwise
    for row in range(1, rows - 1):
        for col in range(1, cols - 1):
            if input[row][col] != "A":
                continue
            corners = (
                input[row - 1][col - 1] +
                input[row - 1][col + 1] +
                input[row + 1][col + 1] +
                input[row + 1][col - 1]
            )
            if corners in valid_corners: 
                count += 1
            
    return count


def main():
    input = open("input.txt", "r").read().splitlines()
    part1_ans = part1(input)
    part2_ans = part2(input)
    print(f"Part 1: {part1_ans}")
    print(f"Part 2: {part2_ans}")

if __name__ == "__main__":
    main()

File: Go/day04/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Point struct {
	x, y int
}

func (p *Point) add(other Point) Point {
	return Point{p.x + other.x, p.y + other.y}
}
func (p *Point) withinBounds(input []string) bool {
	return p.x >= 0 && p.x < len(input) && p.y >= 0 && p.y < len(input[0])
}

func part1(input []string) int {
	count := 0
	dirs := []Point{
		{-1, 0}, {1, 0}, {0, -1}, {0, 1},
		{-1, -1}, {-1, 1}, {1, -1}, {1, 1},
	}

	targets := "MAS"
	for r := range input {
		for c := range input[r] {
			if input[r][c] == 'X' {
				for _, dir := range dirs {
					p := Point{r, c}
					for k := 0; k < 3; k++ {
						p = p.add(dir)
						if !p.withinBounds(input) || input[p.x][p.y] != targets[k] {
							break
						}
						if k == 2 {
							count++
						}
					}
				}
			}
		}
	}

	return count
}

func part2(input []string) int {
	count := 0
	valid := map[string]struct{}{
		"MMSS": {}, "MSSM": {}, "SSMM": {}, "SMMS": {},
	}
	cornerOffsets := []Point{
		{-1, -1}, {-1, 1}, {1, 1}, {1, -1},
	}

	for r := 1; r < len(input)-1; r++ {
		for c := 1; c < len(input[r])-1; c++ {
			if input[r][c] == 'A' {
				corners := make([]byte, len(cornerOffsets))

				for i, offset := range cornerOffsets {
					corners[i] = input[r+offset.x][c+offset.y]
				}

				if _, ok := valid[string(corners)]; ok {
					count++
				}
			}
		}
	}

	return count
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
